pycqed/instrument_drivers/physical_instruments/ZurichInstruments/ZI_HDAWG8.py

```diff
@@ -1,1128 +1,460 @@
-            marker_uint = np.array(marker, dtype=np.uint16)
-            mode += 4
-        
-        if mode == 1:
-            return chan0_uint
-        elif mode == 2:
-            return chan1_uint
-        elif mode == 3:
-            return np.vstack((chan0_uint, chan1_uint)).reshape((-2,),order='F')
-        elif mode == 4:
-            return marker_uint
-        elif mode == 5:
-            return np.vstack((chan0_uint, marker_uint)).reshape((-2,),order='F')
-        elif mode == 6:
-            return np.vstack((chan1_uint, marker_uint)).reshape((-2,),order='F')
-        elif mode == 7:
-            return np.vstack((chan0_uint, chan1_uint, marker_uint)).reshape((-2,),order='F')
-        else:
-            return []
-
-    def upload_waveform_realtime(self, w0, w1, awg_nr: int, wf_nr: int =1):
-        """
-        Warning! This method should be used with care.
-        Uploads a waveform to the awg in realtime, note that this get's
-        overwritten if a new program is uploaded.
+    ##########################################################################
+    # 'public' functions: application specific/codeword support
+    ##########################################################################
 
-        Arguments:
-            w0   (array): waveform for ch0 of the awg pair.
-            w1   (array): waveform for ch1 of the awg pair.
-            awg_nr (int): awg_nr indicating what awg pair to use.
-            wf_nr  (int): waveform in memory to overwrite, default is 1.
-
-        There are a few important notes when using this method
-        - w0 and w1 must be of the same length
-        - any parts of a waveform longer than w0/w1 will not be overwritten.
-        - loading speed depends on the size of w0 and w1 and is ~80ms for 20us.
-
-        """
-        # these two attributes are added for debugging purposes.
-        # they allow checking what the realtime loaded waveforms are.
-        self._realtime_w0 = w0
-        self._realtime_w1 = w1
-
-        # Changed by NielsH (niels@zhinst.com)
-
-        # New dynamic waveform upload
-        c = self._awg_waveform(w0, w1)
-        try:
-            self._dev.daq.setVector('/' + self._devname + 
-                '/awgs/{}/waveform/waves/{}'.format(awg_nr, wf_nr), c)
-        except Exception as e:
-            logging.warning(e) 
-        # Commented out checking if ready. 
-        # creates too much time overhead.
-        # data = self._dev.poll()
-        # t0 = time.time()
-        # while not data: 
-        #     data = self._dev.poll()
-        #     if time.time()-t0> self.timeout(): 
-        #         raise TimeoutError
-        # self._dev.unsubs('awgs/{}/ready'.format(awg_nr))
-        self._dev.seti('awgs/{}/enable'.format(awg_nr), 1)
-        self._dev.sync()
-
-
-
-    def upload_codeword_program(self, awgs=np.arange(4)):
+    def upload_codeword_program(self, awgs=numpy.arange(4), cfg_num_codewords=None, cfg_codeword_protocol=None):
```

Comment:
This method seems completely stripped down. 
I completely miss the part where it is defining what codeword maps to what waveform. 
This is in particular important as the number of codewords differs per operating mode (flux and microwave) and some waveforms are reused. How is this supposed to work now? 

Again, If I am missing something let me know.

Reply:
@AdriaanRol the mapping is done at the time when we decide that we need to upload a new AWG program. At that point, we generate the program consisting of a preamble with all the `setDIO` commands and the small main loop.